Don't close in TSREI in cancelled contexts

werkt opened the pull request:
A cancelled context in a ServerCall will have already closed - this
suppresses log messages indicating "context already closed" from
when SREs are thrown as a result of active client requests within the
context.

thelinuxfoundation commented:
Thank you for your pull request. Before we can look at your contribution, we need to ensure all contributors are covered by a Contributor License Agreement.

After the following items are addressed, please respond with a new comment here, and the automated system will re-verify.

* User @werkt isn't covered by a CLA. They will need to complete the form at https://identity.linuxfoundation.org/projects/cncf

_Regards,
CLA GitHub bot_

werkt commented:
signed cla

zpencer commented:
I'm looking at `Context.addListener` for server side and see only these context cancellation listeners:
https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ServerImpl.java#L511
https://github.com/grpc/grpc-java/blob/master/core/src/main/java/io/grpc/internal/ServerCallImpl.java#L241

I don't see that a cancelled context implies the call already closed. Can you provide more info on how you triggered the warning?

werkt commented:
Sure. Working up a test now, but here are the traces that I encountered (version 1.8.0, so line numbers don't match master)

```
Jun 27, 2018 12:06:29 PM io.grpc.internal.SerializingExecutor run
SEVERE: Exception while executing runnable io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$SerializingServerCall$4@27d4d5c
java.lang.IllegalStateException: call already closed
        at com.google.common.base.Preconditions.checkState(Preconditions.java:501)
        at io.grpc.internal.ServerCallImpl.close(ServerCallImpl.java:166)
        at io.grpc.PartialForwardingServerCall.close(PartialForwardingServerCall.java:46)
        at io.grpc.ForwardingServerCall.close(ForwardingServerCall.java:22)
        at io.grpc.ForwardingServerCall$SimpleForwardingServerCall.close(ForwardingServerCall.java:39)
        at io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$SerializingServerCall.access$301(TransmitStatusRuntimeExceptionInterceptor.java:124)
        at io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$SerializingServerCall$4.run(TransmitStatusRuntimeExceptionInterceptor.java:169)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at com.google.common.util.concurrent.MoreExecutors$DirectExecutor.execute(MoreExecutors.java:397)
        at io.grpc.internal.SerializingExecutor.schedule(SerializingExecutor.java:93)
        at io.grpc.internal.SerializingExecutor.execute(SerializingExecutor.java:86)
        at io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$SerializingServerCall.close(TransmitStatusRuntimeExceptionInterceptor.java:166)
        at io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$1.closeWithException(TransmitStatusRuntimeExceptionInterceptor.java:115)
        at io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$1.onHalfClose(TransmitStatusRuntimeExceptionInterceptor.java:77)
        at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:271)
        at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:648)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
```

```
io.grpc.StatusRuntimeException: DEADLINE_EXCEEDED: context timed out
        at io.grpc.Status.asRuntimeException(Status.java:517)
        at build.buildfarm.instance.stub.Retrier.execute(Retrier.java:204)
        at build.buildfarm.instance.shard.ShardInstance.checkMissingBlob(ShardInstance.java:254)
        at build.buildfarm.instance.shard.ShardInstance.getBlobImpl(ShardInstance.java:417)
        at build.buildfarm.instance.shard.ShardInstance.getBlob(ShardInstance.java:443)
        at build.buildfarm.server.ByteStreamService.readBlob(ByteStreamService.java:60)
        at build.buildfarm.server.ByteStreamService.read(ByteStreamService.java:154)
        at com.google.bytestream.ByteStreamGrpc$MethodHandlers.invoke(ByteStreamGrpc.java:526)
        at io.grpc.stub.ServerCalls$UnaryServerCallHandler$UnaryServerCallListener.onHalfClose(ServerCalls.java:171)
        at io.grpc.PartialForwardingServerCallListener.onHalfClose(PartialForwardingServerCallListener.java:33)
        at io.grpc.ForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:23)
        at io.grpc.ForwardingServerCallListener$SimpleForwardingServerCallListener.onHalfClose(ForwardingServerCallListener.java:40)
        at io.grpc.util.TransmitStatusRuntimeExceptionInterceptor$1.onHalfClose(TransmitStatusRuntimeExceptionInterceptor.java:74)
        at io.grpc.internal.ServerCallImpl$ServerStreamListenerImpl.halfClosed(ServerCallImpl.java:271)
        at io.grpc.internal.ServerImpl$JumpToApplicationThreadServerStreamListener$1HalfClosed.runInContext(ServerImpl.java:648)
        at io.grpc.internal.ContextRunnable.run(ContextRunnable.java:37)
        at io.grpc.internal.SerializingExecutor.run(SerializingExecutor.java:123)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.TimeoutException: context timed out
        at io.grpc.Context$CancellableContext$1.run(Context.java:722)
        at io.netty.util.concurrent.PromiseTask$RunnableAdapter.call(PromiseTask.java:38)
        at io.netty.util.concurrent.ScheduledFutureTask.run(ScheduledFutureTask.java:120)
        at io.netty.util.concurrent.AbstractEventExecutor.safeExecute(AbstractEventExecutor.java:163)
        at io.netty.util.concurrent.SingleThreadEventExecutor.runAllTasks(SingleThreadEventExecutor.java:403)
        at io.netty.channel.nio.NioEventLoop.run(NioEventLoop.java:463)
        at io.netty.util.concurrent.SingleThreadEventExecutor$5.run(SingleThreadEventExecutor.java:858)
        at io.netty.util.concurrent.DefaultThreadFactory$DefaultRunnableDecorator.run(DefaultThreadFactory.java:138)
        ... 1 more
```

Both of these warnings were triggered several calls deep during service method implementations which were making proxied blocking stub calls, which threw an SRE from the stub method call, which, when caught, checked and saw true for Context.current().isCancelled(), and the SRE was rethrown to abandon the call stack.

werkt commented:
I've updated this to include an extension to the TSREI test, and reworked it to observe onCancel and onComplete as 'close-latching' calls on success, as well as handling an exception with a close. This will work if all attempts to close the wrapped server call go through onComplete or onCancel, but not if a close call bypasses this mechanism.

carl-mastrangelo commented on core/src/test/java/io/grpc/util/UtilServerInterceptorsTest.java:
```diff
@@ -111,14 +111,36 @@ public void onReady() {
     final Metadata expectedMetadata;
 
     int numCloses;
+    int numMismatchedCloses;
```
did you mean to read this?

werkt replied:
I did not. Thanks for the catch.

zpencer commented:
Should the `closeCalled` be tracked in the serializing server call, in case close was called on it directly?

ejona86 commented:
@zpencer... probably. I'm going to wait until @zpencer approves.

zpencer commented on core/src/main/java/io/grpc/util/TransmitStatusRuntimeExceptionInterceptor.java:
```diff
@@ -160,12 +161,16 @@ public void run() {
+        serializingExecutor.execute(new Runnable() {
+          @Override
+          public void run() {
+            SerializingServerCall.super.close(status, trailers);
```
`closeCalled` can only be safely read/written from this runnable, which serializes the access in a thread safe way

zpencer commented on core/src/test/java/io/grpc/util/UtilServerInterceptorsTest.java:
```diff
@@ -111,7 +116,29 @@ public void onReady() {
     getSoleMethod(intercepted).getServerCallHandler().startCall(call, headers).onHalfClose();
     getSoleMethod(intercepted).getServerCallHandler().startCall(call, headers).onReady();
-    assertEquals(5, call.numCloses);
+    ServerCall.Listener<Void> callDoubleSreListener =
```
This test method is starting to test a lot of different conditions. I would prefer if this new test code is added as two new test cases.

zpencer commented:
Sorry for the delay @werkt , thank you for your PR!

ejona86 commented:
@zpencer, would this be good to back-port to 1.14?

zpencer commented:
Yes, this would be a good thing to patch. I'll submit a PR.
